source/slang/slang-check-decl.cpp

```diff
@@ -7578,8 +7571,49 @@ bool SemanticsVisitor::findWitnessForInterfaceRequirement(
     // and if nothing is found we print the candidates that made it
     // furthest in checking.
     //
-    if (!specificDiagnosticEmitted)
+    // Based on the failure reason, emit specific diagnostics
+    if (failureDetails.reason == WitnessSynthesisFailureReason::MethodResultTypeMismatch)
     {
+        // Emit specific return type mismatch diagnostic
+        getSink()->diagnose(
+            failureDetails.candidateMethod,
+            Diagnostics::memberReturnTypeMismatch,
+            failureDetails.candidateMethod,
+            failureDetails.actualType,
+            failureDetails.expectedType);
+            
+        getSink()->diagnose(
+            failureDetails.requiredMethod,
+            Diagnostics::seeDeclarationOfInterfaceRequirement,
+            failureDetails.requiredMethod);
+    }
+    else if (failureDetails.reason == WitnessSynthesisFailureReason::MethodParameterMismatch)
+    {
+        // Emit specific parameter mismatch diagnostic
```

Comment:
This branch is implemented exactly the same as the general case below, maybe just remove it and the `WitnessSynthesisFailureReason::MethodParameterMismatch` enum case for now.